crates/consensus/src/engine/mod.rs

```diff
@@ -322,3 +326,420 @@ where
+                rx: UnboundedReceiverStream::new(rx),
+            };
+
+            tokio::spawn(engine);
+
+            let (result_tx, result_rx) = oneshot::channel();
+            let parent = transform_block(random_block(100, None, None, Some(0)), |mut b| {
+                b.header.difficulty = config.merge_terminal_total_difficulty.into();
+                b
+            });
+            let block = random_block(101, Some(parent.hash()), None, Some(0));
+
+            client.add_block(parent.hash(), parent.clone().unseal());
+
+            tx.send(EngineMessage::NewPayload(block.clone().into(), result_tx))
+                .expect("failed to send engine msg");
+
+            let result = result_rx.await;
+            assert_matches!(result, Ok(Ok(_)));
+            let expected_result = PayloadStatus::from_status(PayloadStatusEnum::Invalid {
+                validation_error: EngineApiError::PayloadPreMerge.to_string(),
+            });
+            assert_eq!(result.unwrap().unwrap(), expected_result);
+        }
+
+        #[tokio::test]
+        async fn invalid_payload_timestamp() {
+            let (tx, rx) = unbounded_channel();
+            let config = Config::default();
+            let client = Arc::new(MockEthProvider::default());
+            let engine = EthConsensusEngine {
+                client: client.clone(),
+                config: config.clone(),
+                local_store: Default::default(),
+                rx: UnboundedReceiverStream::new(rx),
+            };
+
+            tokio::spawn(engine);
+
+            let (result_tx, result_rx) = oneshot::channel();
+            let block_timestamp = 100;
+            let parent_timestamp = block_timestamp + 10;
+            let parent = transform_block(random_block(100, None, None, Some(0)), |mut b| {
+                b.header.timestamp = parent_timestamp;
+                b.header.difficulty = (config.merge_terminal_total_difficulty + 1).into();
+                b
+            });
+            let block =
+                transform_block(random_block(101, Some(parent.hash()), None, Some(0)), |mut b| {
+                    b.header.timestamp = block_timestamp;
+                    b
+                });
+
+            client.add_block(parent.hash(), parent.clone().unseal());
+
+            tx.send(EngineMessage::NewPayload(block.clone().into(), result_tx))
+                .expect("failed to send engine msg");
+
+            let result = result_rx.await;
+            assert_matches!(result, Ok(Ok(_)));
+            let expected_result = PayloadStatus::from_status(PayloadStatusEnum::Invalid {
+                validation_error: EngineApiError::PayloadTimestamp {
+                    invalid: block_timestamp,
+                    latest: parent_timestamp,
+                }
+                .to_string(),
+            });
+            assert_eq!(result.unwrap().unwrap(), expected_result);
+        }
+
+        // TODO: add execution tests
```

Comment:
Wdym?

Reply:
these tests cover everything until the actual payload execution, i'll add it separately since it there is a lot of state preparation and block construction needed